### PR TITLE
Upgrade to Hudi 0.9.0. Integrate with Hudi's Metadata based listing and Bootstrap Features

### PR DESCRIPTION
**Pull request by umehrot2**
Upgrade to Hudi 0.9.0. Integrate with Hudi's Metadata based listing and Bootstrap Features

Test plan - (Please fill in how you tested your changes)

Please make sure your submission complies with our [Development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [Formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), and [Commit Message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests) guidelines. Don't forget to follow our [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution) for any code copied from other projects.

Fill in the release notes towards the bottom of the PR description.
See [Release Notes Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) for details.

```
== RELEASE NOTES ==

General Changes
* ...
* ...

Hive Changes
* ...
* ...
```

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by n3nash** [comment]
@vkorukanti Could you please help review this ?

**Comment by umehrot2** [comment]
Can someone help review this ?

**Comment by vinothchandar** [comment]
Hi all, could we get some eyes on this please?  @zhenxiao could you help out or connect us with someone who can help us. ?

**Comment by dborkar** [comment]
@imjalpreet @ashishtadose  @beinan This is an important PR for the Presto + Hudi integration. can one of you take a look? 

@aweisberg FYI.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java [hunk 1]
```diff
@@ -357,13 +365,85 @@ static String getInputFormatName(Properties schema)
...
+     *
+     * @param configuration Hadoop configuration
+     * @param storage Hive storage information
+     * @param customSplitInfo Custom split information
+     * @return Boolean indicating whether to use RecordReader or not.
+     */
+    public static boolean shouldUseRecordReaderFromInputFormat(Configuration configuration, Storage storage,
+                                                               Map<String, String> customSplitInfo)
+    {
+        if (customSplitInfo != null && customSplitInfo.containsKey(HiveUtil.CUSTOM_FILE_SPLIT_CLASS_KEY)) {
```
Comment (beinan) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:381]: static import HiveUtil.CUSTOM_FILE_SPLIT_CLASS_KEY

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java [hunk 2]
```diff
@@ -357,13 +365,85 @@ static String getInputFormatName(Properties schema)
...
+            InputFormat<?, ?> inputFormat = HiveUtil.getInputFormat(configuration,
+                    storage.getStorageFormat().getInputFormat(), false);
+            return Arrays.stream(inputFormat.getClass().getAnnotations())
+                    .map(Annotation::annotationType)
+                    .map(Class::getSimpleName)
+                    .anyMatch(name -> name.equals("UseRecordReaderFromInputFormat"));
```
Comment (beinan) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:387]: make "UseRecordReaderFromInputFormat" as a constant?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java [hunk 3]
```diff
@@ -357,13 +365,85 @@ static String getInputFormatName(Properties schema)
...
+     * @param conf Hadoop Configuration
+     * @param tablePath Base path for the table
+     * @return Boolean indicating whether to use File Splitting logic from Input Format or not.
+     */
+    static boolean shouldUseFileSplitsForHudi(ConnectorSession session, InputFormat<?, ?> inputFormat, Configuration conf,
+                                              String tablePath)
+    {
+        // Set Hudi metadata based listing configurations to be used through input format
+        conf.set(HoodieMetadataConfig.METADATA_ENABLE_PROP, String.valueOf(isPreferMetadataToListHudiFiles(session)));
+        conf.set(HoodieMetadataConfig.METADATA_VALIDATE_PROP, String.valueOf(isHudiMetadataVerificationEnabled(session)));
```
Comment (beinan) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:429]: static import HoodieMetadataConfig.METADATA_ENABLE_PROP and HoodieMetadataConfig.METADATA_VALIDATE_PROP

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java [hunk 4]
```diff
@@ -357,13 +365,85 @@ static String getInputFormatName(Properties schema)
...
+     * @param tablePath  Base path for the table
+     * @return Boolean indicating whether to use File Splitting logic from Input Format or not.
+     */
+    static boolean shouldUseFileSplitsFromInputFormat(ConnectorSession session, InputFormat<?, ?> inputFormat,
+                                                      Configuration conf, String tablePath)
+    {
+        boolean hasUseSplitsAnnotation = Arrays.stream(inputFormat.getClass().getAnnotations())
+                .map(Annotation::annotationType)
+                .map(Class::getSimpleName)
+                .anyMatch(name -> name.equals("UseFileSplitsFromInputFormat"));
```
Comment (beinan) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java:408]: Make UseFileSplitsFromInputFormat as a constant ?

File: presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        public boolean hasNext()
+        {
+            return hoodieBaseFileIterator.hasNext();
+        }
+
+        @Override
+        public HiveFileInfo next() throws IOException
+        {
+            FileStatus fileStatus = hoodieBaseFileIterator.next().getFileStatus();
+            String[] name = new String[]{"localhost:" + DFSConfigKeys.DFS_DATANODE_DEFAULT_PORT};
```
Comment (beinan) [inline presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java:108]: static import DFSConfigKeys.DFS_DATANODE_DEFAULT_PORT

**Review by beinan — COMMENTED** [review]
looks good to me, thank you for the contribution!  just a couple of coding style issues.

**Comment by vinothchandar** [comment]
Thanks @dborkar and @beinan ! 

@umehrot2 should we also upgrade to 0.8.0 given it has already been out with some fixes around metadata table as well?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java (hunk 1 above)
Comment (umehrot2) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java]: Done.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java (hunk 2 above)
Comment (umehrot2) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java]: Done.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java (hunk 3 above)
Comment (umehrot2) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java]: Done.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java (hunk 4 above)
Comment (umehrot2) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java]: Done.

File: presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java (hunk 5 above)
Comment (umehrot2) [reply presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java]: Done.

**Comment by umehrot2** [comment]
@beinan sorry for the delay here. I have addressed your comments, resolved conflicts and upgraded to the latest Hudi release i.e 0.9.0. Can you please take another look ? Also, it seems the workflow is not running and requires a maintainer to trigger it.

**Comment by rubenssoto** [comment]
hi @umehrot2 

I would like to try metadata listing but, I tried to compile your branch and I had some errors:
```
ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.0.0:check (checkstyle) on project presto-hive: You have 2 Checkstyle violations. -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :presto-hive
```

File: presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+import static com.facebook.presto.hive.HiveFileInfo.createHiveFileInfo;
+import static com.facebook.presto.hive.HiveSessionProperties.isHudiMetadataVerificationEnabled;
+import static com.facebook.presto.hive.HiveSessionProperties.isPreferMetadataToListHudiFiles;
+import static org.apache.hadoop.hdfs.DFSConfigKeys.DFS_DATANODE_DEFAULT_PORT;
+
+public final class HudiDirectoryLister
+        implements DirectoryLister
+{
+    private static final Logger LOGGER = Logger.get(HudiDirectoryLister.class);
```
Comment (beinan) [inline presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java:50]: I know we have mixed naming convention for logger,  but I think we're using lower case "log" for the instance of Logger.

File: presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        public boolean hasNext()
+        {
+            return hoodieBaseFileIterator.hasNext();
+        }
+
+        @Override
+        public HiveFileInfo next() throws IOException
+        {
+            FileStatus fileStatus = hoodieBaseFileIterator.next().getFileStatus();
+            String[] name = new String[]{"localhost:" + DFS_DATANODE_DEFAULT_PORT};
```
Comment (beinan) [inline presto-hive/src/main/java/com/facebook/presto/hive/HudiDirectoryLister.java:112]: Looks like you hardcoded "localhost" here,  is it really you intend to do?

**Review by beinan — COMMENTED** [review]
Thank you @umehrot2 !  It looks good to me except the hardcoded "localhost",  could you help to provide more context about why you're using a hardcoded "localhost" in your code?  Thanks!

**Comment by vinothchandar** [comment]
@arunthirupathi thanks for the feedback. I think that makes sense. We can upgrade and solve any bundling/packaging first. then land the changes. 

@umehrot2 please let me know what you think. cc @codope as well

**Comment by codope** [comment]
> I went through the changes, looks like it has lot of breaking changes.
> 
> 1. Some caches are removed.
> 2. Previously two class types were supported for Hudi, now only one class.
> 
> It would be great, if it can be organized into smaller changes, so that they can be reviewed easily.

@arunthirupathi Thanks for the review. I am going to break it down into 3 PRs:
1. Just the version upgrade.
2. Hudi bootstrap related changes.
3. Integrate with Hudi's metadata listing.
Will ensure that the new PRs address the comments on this one.

Meanwhile, could you please review the upgrade PR #16881 ? The other changes depend on this upgrade.

**Pull request closed without merging**